portfolio/src/main/java/com/google/sps/servlets/LikeServlet.java

```diff
@@ -0,0 +1,68 @@
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import java.util.ConcurrentModificationException;
+import com.google.appengine.api.datastore.Transaction;
+
+/* 
+REQUEST: a POST query string like this "/like?likeID=1234567890123456"
+Behavior: update "like" counter in the database
+*/
+
+@WebServlet("/like")
+public class LikeServlet extends HttpServlet {
+  private DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {    
+    long commentId;
+    try {
+      commentId = Long.parseLong(request.getParameter("commentId"));
+    } catch (NumberFormatException e) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Invalid comment ID.");
+      return;
+    }
+    Key likeKey = KeyFactory.createKey("Comment", commentId);
+    // Wrap retrieving and storing entity in one transaction
+    // to avoid counting collision. 
+    int retries = 3;
+    while (true) {
+      Transaction txn = datastore.beginTransaction();
+      Entity commentEntity;
+      try {
+        commentEntity = datastore.get(likeKey);        
+      } catch (EntityNotFoundException e) {
+        response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Comment not found.");
+        break;
+      }
+      long count = (long) commentEntity.getProperty("likeCount");
+      ++count;
+      commentEntity.setProperty("likeCount", count);
+      datastore.put(txn, commentEntity);
+      try {
+        txn.commit();
+        break;
+      } catch (ConcurrentModificationException e) {
+        if (retries == 0) {
+          response.sendError(HttpServletResponse.SC_CONFLICT, "Database busy.");
+        }
+        --retries;
+      } finally {
+        if (txn.isActive()) {
+          txn.rollback();
+        }
+      }
+    } // end of while loop
```

Comment:
Last request!
This method is getting pretty dense. Would be a good idea to use some helper methods. Check out Jennifer's PR on an idea on how to handle this.

Look around this code block.
https://github.com/nif33/my-portfolio/pull/9/files#r438338566

Reply:
Haha no worries any request is welcomed! 

I refactored the code and kept
1. `datastore.beginTransaction()` and `txn.commit()` 
2. all `response.sendError(..)` 

in the while loop but outside the helper function, because this way it is easier to identify the lifetime of the transaction and to identify the error sources. Is this what was requested?

Reply:
Hm.. I was looking more for a function that does the read/write transactions and the caller would deal with the retries while also removing the seemingly infinite while loop.

Something like
```
boolean isWriteSuccessful = false;
for (int retries = 0; retries < MAX_RETRIES; retries++) {
  try {
    incrementCommentLikeCounter(commentKey);
  } catch (ConcurrentModificationException e) {
    continue;
  } catch (EntityNotFoundException e) {
    response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Comment not found.");
     return;
  }
  isWriteSuccessful = true;
  break;
}

if (!isWriteSuccessful) {
  response.sendError(HttpServletResponse.SC_CONFLICT, "Database busy.");
  return;
}
```

Notice how there isn't more than two indentation levels. Notice how it easier to understand from the first few lines that this loop is trying to successfully increment the like counter  with MAX_RETRIES max attempts.

Since the transaction logic is hidden inside the method, you would need to rollback the transaction inside of it as well.

```
private void incrementCommentLikeCounter(Key commentKey) {
  Transaction txn = datastore.beginTransaction();
  try {
    Entity commentEntity = datastore.get(txn, commentKey);
    long likeCount = (long) commentEntity.getProperty("likeCount");
    ++likeCount;
    commentEntity.setProperty("likeCount", likeCount);
    datastore.put(txn, commentEntity);
    txn.commit();
  } catch (ConcurrentModificationException e) {
    throw e;
  } catch (EntityNotFoundException e) {
    throw e;
  } finally {
    if (txn.isActive()) {
      txn.rollback();
    }
  }
} 
```

Note: Notice how I also put txn in the get. Don't forget that :) Then in this case only the write is happening in a transaction.

You can also see I sacrificed repetitive try catch blocks to make it easier to read on a per-method level. This probably can be cleaner but we shouldn't spend too much time on this!

Reply:
Done. Thank you for literally rewriting this block for me lol. I see how you hide the transactions inside the helper function and use a for-loop instead of a while-loop to make retries more obvious. 

About putting txn in get, the[ sample code](https://cloud.google.com/appengine/docs/standard/java/datastore/transactions#uses_for_transactions) they have omit the txn for some reason, maybe a bug?

Reply:
Maybe. I'm not completely sure. Maybe because its not bad to read before or after someone has already written.

Cases for when you only do txn on write:
(1) Someone else reads then writes and then you finally read: No issue you read the most updated value and its all good
(2) Someone else reads then you read then they write then you write: Your commit would throw the exception because a write from someone else was done after you began your txn. 

There are probably more cases but it seems like they would all work as you intend it. I just think its cleaner that you group the read and write into 1 transaction. Its like saying "EVERYONE STOP WHAT YOU ARE DOING!" and you proceed to read and write without anyone else trying to get in your way to read as well.

Reply:
I tested both by inserting thread.sleep between get and put and it works in both cases (have txn in get or not) so it is not a bug. And I think you are right, that all possible cases do work out well. It only throws `ConcurrentModificationException` when someone writes after you have read but before you write anything. 

I also agree that bundling read and write together into the same txn is more readable and avoids the ambiguity. It might be slower than otherwise but in this application that's quite negligible.